app/src/main/java/org/android/bbangzip/presentation/ui/todo/TodoViewModel.kt

```diff
@@ -0,0 +1,358 @@
+
+                is TodoEvent.OnListItemMove -> {
+                    val from = event.from
+                    val to = event.to
+
+                    if (from == to) return
+
+                    val currentFlatList = currentUiState.flatList.toMutableList()
+                    val movedItem = currentFlatList.removeAt(from)
+                    currentFlatList.add(to, movedItem)
+
+                    synchronizeListState(currentFlatList)
+                    val newCategories = reconstructCategoriesFromFlatList(currentFlatList)
+
+                    updateState(UpdateCategories(newCategories))
+                    updateState(UpdateFlatList(currentFlatList.toList()))
+                }
+
+                is TodoEvent.OnMenuClick -> {
+                    updateState(UpdateIsMenuOpen(isMenuOpen = !currentUiState.isMenuOpen))
+                }
+
+                TodoEvent.OnAddCategoryClick -> TODO()
+                TodoEvent.OnCategoryChipClick -> TODO()
+                TodoEvent.OnCommitmentAreaClick -> TODO()
+                TodoEvent.OnDateChanged -> TODO()
+                TodoEvent.OnManageCategoryClick -> TODO()
+            }
+        }
+
+        override fun reduceState(
+            state: TodoState,
+            reduce: TodoReduce,
+        ): TodoState {
+            return when (reduce) {
+                is UpdateLoading -> state.copy(isLoading = reduce.isLoading)
+                is UpdateMotivationMessage -> state.copy(motivationMessage = reduce.message)
+                is UpdateCategories -> state.copy(categories = reduce.categories)
+                is UpdateFlatList -> state.copy(flatList = reduce.flatList)
+                is UpdateIsMenuOpen -> state.copy(isMenuOpen = reduce.isMenuOpen)
+                is UpdateSelectedDate -> state.copy(selectedDate = reduce.selectedDate)
+            }
+        }
+
+        private fun updateCategoriesAndFlatList(categories: List<Category>) {
+            updateState(UpdateCategories(categories))
+            val flatList = categories.toFlatList()
+            updateState(UpdateFlatList(flatList))
+        }
+
+        private fun List<Category>.toFlatList(): List<ListItem> {
+            return this.flatMap { category ->
+                val categoryItem = ListItem.CategoryItem(category)
+                val todoItems =
+                    category.todos.mapIndexed { index, todo ->
+                        ListItem.TodoItem(
+                            todo = todo,
+                            category = category,
+                            isLastInCategory = index == category.todos.size - 1,
+                        )
+                    }
+                listOf(categoryItem) + todoItems
+            }
+        }
+
+        private fun synchronizeListState(
+            list: MutableList<ListItem>,
+        ) {
+            var currentCategory: ListItem.CategoryItem? = null
+
+            for (i in list.indices) {
+                when (val item = list[i]) {
```

Comment:
```suggestion
            for (i in list.indices) { i , item ->
                when (item) {
```
이건 어떤가요?

Reply:
```suggestion
            for (item in list) {
                when (item) {
```
이렇게 말씀하시는거맞죠? 하지만 인덱스가 필요해서 고칠거면 for문 말고 아예 forEachIndexed를 써야할 것 같습니다.

Reply:
네! 선언하는것보다 가독성이 높아지는것 같아요